hal/inc/ble_hal.h

```diff
@@ -15,190 +15,747 @@
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_delete_whitelist(void* reserved);
 
-int ble_start_advert(void* reserved);
-void ble_stop_advert(void* reserved);
+/**
+ * Set the TX Power for advertisement.
+ *
+ * @param[in]   value   Radio transmit power in dBm.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_set_tx_power(int8_t value);
 
-int ble_get_char_param(uint16_t conn_handle, uint16_t char_handle, ble_char_param* param, void* reserved);
-int ble_set_char_value(uint16_t conn_handle, uint16_t char_handle, const char* data, uint16_t size, unsigned flags, void* reserved);
+/**
+ * Get the TX Power for advertisement.
+ *
+ * @returns     The current TX power.
+ */
+int8_t ble_gap_get_tx_power(void);
+
+/**
+ * Set the BLE advertising parameters. Changing the advertising parameter during
+ * advertising will restart advertising using the new parameters.
+ *
+ * @param[in]   adv_params  Pointer to the advertising parameters to be set.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_set_advertising_parameters(const hal_ble_adv_params_t* adv_params, void* reserved);
+
+
+/**
+ * Get the current BLE advertising parameters.
+ *
+ * @param[in,out]   adv_params  Pointer to the advertising parameters to be filled.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_get_advertising_parameters(hal_ble_adv_params_t* adv_params, void* reserved);
+
+/**
+ * Set the BLE advertising data. It will update the advertising data immediately if success.
+ * Changing the advertising data during advertising will restart advertising using the
+ * previous advertising parameters and the new advertising data.
+ *
+ * @param[in]   buf     Pointer to the advertising data to be set.
+ * @param[in]   len     Length of the advertising data.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_set_advertising_data(const uint8_t* buf, uint16_t len, void* reserved);
+
+/**
+ * Get the current BLE advertising data.
+ *
+ * @param[in,out]   buf     Pointer to the advertising data to be filled.
+ * @param[in]       len     Length of the given buffer.
+ *
+ * @returns     The length of copied data.
+ */
+size_t ble_gap_get_advertising_data(uint8_t* buf, uint16_t len, void* reserved);
+
+/**
+ * Set the BLE scan response data.
+ *
+ * @param[in]   buf     Pointer to the scan response data to be set.
+ * @param[in]   len     Length of the scan response data.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_set_scan_response_data(const uint8_t* buf, uint16_t len, void* reserved);
+
+/**
+ * Get the current BLE scan response data.
+ *
+ * @param[in,out]   buf     Pointer to the scan response data to be filled.
+ * @param[in]       len     Length of the given buffer.
+ *
+ * @returns     The length of copied data.
+ */
+size_t ble_gap_get_scan_response_data(uint8_t* buf, uint16_t len, void* reserved);
```

Comment:
Consider returning `ssize_t` or `int` to possibly return a `system_error_t`.

If returning `size_t`, perhaps incoming `len` should also be `size_t`?